### PR TITLE
Use a new RPC to get evaluated Service objects

### PR DESCRIPTION
**Pull request by tgjonestx**
Fixes CC-2591

Replace recursive RPC calls for GetService from delegate to master with a single GetEvaluatedService RPC which runs the template evaluations on the master.

### CODE REVIEW

File: dao/elasticsearch/service.go [hunk 1]
```diff
@@ -128,6 +128,15 @@ func (this *ControlPlaneDao) GetService(id string, myService *service.Service) e
 	return err
 }
 
+// GetEvaluatedService returns a service where an evaluation has been executed against all templated properties.
```
Comment (jplouis) [inline dao/elasticsearch/service.go]: Please don't add methods to the dao package.

File: facade/service.go [hunk 2]
```diff
@@ -733,6 +733,47 @@ func (f *Facade) GetService(ctx datastore.Context, id string) (*service.Service,
 	return svc, nil
 }
 
+// GetEvaluatedService returns a service where an evaluation has been executed against all templated properties.
+func (f *Facade) GetEvaluatedService(ctx datastore.Context, serviceID string, instanceID int) (*service.Service, error) {
+	glog.V(3).Infof("Facade.GetEvaluatedService: serviceID=%s instanceID=%d", serviceID, instanceID)
```
Comment (jplouis) [inline facade/service.go]: Should this use new logging?

File: node/agent.go [hunk 3]
```diff
@@ -280,13 +261,18 @@ func manageTransparentProxy(endpoint *service.ServiceEndpoint, addressConfig *ad
...
 
-	// Evaluate service template fields
-	if err := a.evaluateService(client, svc, instanceID); err != nil {
-		glog.Errorf("Could not evaluate service %s (%s) templates: %s", svc.Name, svc.ID, err)
+	request := dao.EvaluateServiceRequest{
+		ServiceID:  svc.ID,
+		InstanceID: instanceID,
+	}
+	if err := client.GetEvaluatedService(request, svc); err != nil {
+		glog.Errorf("Could not get service %s (%s): %s", svc.Name, svc.ID, err)
```
Comment (jplouis) [inline node/agent.go]: New logging?

**Comment by tgjonestx** [comment]
W.r.t. node agent using dao.ControlPlane, the last part that needs to be ported to rpc/master are the 4 methods in dao/elasticsearch/health.go

**Comment by smousa** [comment]
Make sure to coordinate with @morr-zenoss

File: cli/api/daemon.go [hunk 4]
```diff
@@ -876,7 +878,7 @@ func initMetricsClient() *metrics.Client {
 	return client
 }
 
-func (d *daemon) initFacade() *facade.Facade {
+func (d *daemon) initFacade() (*facade.Facade, error) {
```
Comment (smousa) [inline cli/api/daemon.go]: Does it make sense for this method to call fatalf?

File: node/agent_proxy.go [hunk 5]
```diff
@@ -158,55 +123,10 @@ func (a *HostAgent) AckProxySnapshotQuiece(snapshotId string, unused *interface{
...
-		return err
-	}
-	defer controlClient.Close()
-	err = controlClient.LogHealthCheck(result, unused)
-	return err
-}
 
 // ReportHealthStatus proxies ReportHealthStatus to the master server.
 func (a *HostAgent) ReportHealthStatus(req dao.HealthStatusRequest, unused *int) error {
+	// FIXME: use new master.ClientInterface instead
```
Comment (smousa) [inline node/agent_proxy.go]: Switch over?

File: node/agent_proxy.go [hunk 6]
```diff
@@ -218,6 +138,7 @@ func (a *HostAgent) ReportHealthStatus(req dao.HealthStatusRequest, unused *int)
 
 // ReportInstanceDead proxies ReportInstanceDead to the master server.
 func (a *HostAgent) ReportInstanceDead(req dao.ServiceInstanceRequest, unused *int) error {
+	// FIXME: use new master.ClientInterface instead
```
Comment (smousa) [inline node/agent_proxy.go]: Switch over?

File: node/container.go [hunk 7]
```diff
@@ -123,19 +124,17 @@ func (a *HostAgent) StartContainer(cancel <-chan interface{}, svc *service.Servi
 		logger.WithError(err).Debug("Could not pull the service image")
 		return nil, nil, err
 	}
-	svc.ImageID = imageName
 
 	// Establish a connection to the master
-	// TODO: use the new rpc calls instead
-	client, err := NewControlClient(a.master)
+	masterClient, err := master.NewClient(a.master)
```
Comment (smousa) [inline node/container.go]: Check with @morr-zenoss for changes to this file

File: rpc/master/service_client.go [hunk 8]
```diff
@@ -35,6 +35,11 @@ type WaitServiceRequest struct {
 	Recursive  bool
 }
 
+type EvaluateServiceRequest struct {
```
Comment (smousa) [inline rpc/master/service_client.go]: this I think should go into service_server.go

Comment (smousa) [inline rpc/master/service_client.go]: I think these requests objects make more sense in the *_server.go file:
see https://github.com/control-center/serviced/blob/12b81313b3eec2f3d6353a6ec6733299a2a5fcb6/rpc/master/instance_server.go#L32-L41

File: node/agent_proxy.go (hunk 5 above)
Comment (tgjonestx) [inline node/agent_proxy.go]: Yes - I'm in the process of moving all of `dao/elasticsearch/health.go` stuff to `rpc`

File: node/agent_proxy.go (hunk 6 above)
Comment (tgjonestx) [inline node/agent_proxy.go]: Yes - I'm in the process of moving all of `dao/elasticsearch/health.go` stuff to `rpc`

File: cli/api/daemon.go (hunk 4 above)
Comment (tgjonestx) [inline cli/api/daemon.go]: Well, I was following the pattern that seemed most common for the other initialization calls which was to let the caller, `startMaster`, issue the `Fatal()` call.

File: rpc/master/service_client.go (hunk 8 above)
Comment (tgjonestx) [inline rpc/master/service_client.go]: Agreed - will move

File: rpc/master/service_server.go [hunk 9]
```diff
@@ -13,7 +13,31 @@
...
+}
+
+type WaitServiceRequest struct {
+	ServiceIDs []string
+	State      service.DesiredState
+	Timeout    time.Duration
+	Recursive  bool
+}
+
+type EvaluateServiceRequest struct {
```
Comment (smousa) [inline rpc/master/service_server.go]: Is this the same as ServiceInstance request?

Comment (tgjonestx) [inline rpc/master/service_server.go]: Yes.

I guess I waffled on that a bit. Should we have unique xyzRequest object for each request (even ones that have the same inputs), or should we share a common object for different calls which need the same arguments. When I created EvaluateServiceRequest I was thinking the former, but I guess I got a little lazy as I got deeper into the refactor and opted for the later.

Thoughts?

**Pull request merged**